File: scratch/check_json.php
<?php
$lines = explode("\n", file_get_contents(__DIR__ . '/../public/swagger.json'));

for ($i = 1; $i <= count($lines); $i++) {
    $slice = implode("\n", array_slice($lines, 0, $i));
    // Check if syntax error happens at line $i
    // We try to close open braces to see if valid
    $openBraces = substr_count($slice, '{') - substr_count($slice, '}');
    $openBrackets = substr_count($slice, '[') - substr_count($slice, ']');
    $testJson = $slice . str_repeat('}', max(0, $openBraces)) . str_repeat(']', max(0, $openBrackets));
    
    // Check for trailing commas
    $cleaned = preg_replace('/,\s*([}\]])/', '$1', $testJson);
    json_decode($cleaned);
    if (json_last_error() === JSON_ERROR_SYNTAX) {
        echo "Syntax error first appears around line {$i}: " . trim($lines[$i-1]) . "\n";
        break;
    }
}

File: scratch/check_json.py
import json

try:
    with open('public/swagger.json', 'r', encoding='utf-8') as f:
        data = json.load(f)
    print("VALID JSON!")
except json.JSONDecodeError as e:
    print(f"JSON ERROR: {e.msg} at line {e.lineno}, col {e.colno}, pos {e.pos}")
    with open('public/swagger.json', 'r', encoding='utf-8') as f:
        lines = f.readlines()
        start = max(0, e.lineno - 5)
        end = min(len(lines), e.lineno + 5)
        for i in range(start, end):
            print(f"{i+1}: {lines[i]}", end='')
